Enable providers to do signatures, and implement DSA in the default provider

mattcaswell opened the pull request:
This makes EVP_PKEY_sign and EVP_PKEY_sign_init provider aware. It
also introduces the new type EVP_SIGNATURE to represent signature
algorithms. This also automatically makes the EVP_Sign* APIs provider
aware because they use EVP_Digest* (which is already provider aware)
and EVP_PKEY_sign(_init) under the covers.

I've not yet attempted to untangle the logic in EVP_DigestSign*. That will need to be a different PR.

Finally I've moved DSA handing into the default provider

mattcaswell commented:
This WIP, mainly because it is missing documentation.

levitte commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -300,6 +304,7 @@ static int dsa_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
         return 0;
     BN_free(to->pkey.dsa->g);
     to->pkey.dsa->g = a;
+    to->pkey.dsa->dirty_cnt++;
```
Ah, you caught one that I missed in #9670, cool

levitte commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -514,6 +521,46 @@ static int dsa_pkey_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
+    OSSL_PARAM *params;
+    void *provkey = NULL;
+
+    if (p == NULL || q == NULL || g == NULL || pub_key == NULL)
```
Not `pub_key`, as this may be just the domain parameters, which should be perfectly ok.
See https://github.com/openssl/openssl/pull/9670/files#diff-cd5a6dfdd711380eaa038ea0d9a0959dR540-R544

mattcaswell replied:
Fixed.

levitte commented on crypto/evp/evp_locl.h:
```diff
@@ -108,6 +108,22 @@ struct evp_keyexch_st {
+
+    OSSL_OP_signature_newctx_fn *newctx;
+    OSSL_OP_signature_sign_init_fn *sign_init;
+    OSSL_OP_signature_sign_fn *sign;
```
I'm surprised there aren't any verification functions.  After all, the name `EVP_SIGNATURE` suggests that this method operates on signatures, so it would be *weird* if verification functionality is on another type.  Can I assume that it's upcoming functionality in this method?

mattcaswell replied:
Yes, I expect verification to be part of the same operation. I had been focusing on the sign aspects - but I should probably add verify to this PR too.

mattcaswell replied:
Fixed

levitte commented on include/openssl/core_names.h:
```diff
@@ -101,10 +101,22 @@ extern "C" {
+/* DSA Parameters */
+#define OSSL_PKEY_PARAM_DSA_P        "dsa-p"
+#define OSSL_PKEY_PARAM_DSA_G        "dsa-g"
+#define OSSL_PKEY_PARAM_DSA_Q        "dsa-q"
```
@slontis suggested that these and the DH parameters could be made the same.  See how that ended up in #9670

mattcaswell replied:
Fixed

levitte commented:
Apart from a few details, the DSA changes here are the same or better than in #9670.  This PR being more complete with regards to DSA, I'll rather work with them here than there, and a future rebase will have them dropped in #9670.

mattcaswell commented:
I've updated this PR following the above feedback, and added the missing documentation. Also rebased to pick up the fixes from #9678.

Now out of WIP.

paulidale commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -161,6 +161,17 @@ up to the size of the DH prime B<p>.
+
+Sets the output size of the digest algorithm used for the input to the signature
+functions.
+Internal algorithms that support this parameter are DSA.
```
`The internal algorithm that supports this...`

If there is only one, it should be singular.

mattcaswell replied:
I actually wrote it this way on purpose because I knew we were going to add more in the future. But for now I've written it as you suggested.

slontis commented on crypto/cms/cms_sd.c:
```diff
@@ -706,11 +706,23 @@ int CMS_SignerInfo_sign(CMS_SignerInfo *si)
         si->pctx = pctx;
     }
 
+    /*
```
That appears to be true in the source code..
I have thought the same thing about some other controls only to find they were used by external engines? Could that be possible?

mattcaswell replied:
Yes, its possible although I couldn't find any evidence of that. There was a case of this with the MICALG ctrl which was not used internally, but *was* used by the GOST engine (remembering of course that the GOST engine *used* to be part of our source). But I specifically checked GOST, and its not used there either. I think it is quite unlikely to be used by engines implementing algorithms which we also implement. If we implement those algorithms and didn't need those ctrls, I don't see why an external engine would need them. So, this is perhaps more likely for algorithms implemented by engines that we don't have internally. This seems to be much rarer - GOST is the only example I can think of.
So, while it is possible and I can't rule it out, I think it is quite unlikely.

levitte replied:
Heh, I just looked back at 1.0.2, where `engines/ccgost` still exists, and that version of it doesn't use MICALG at all...

mattcaswell replied:
Interesting

slontis commented on crypto/evp/evp_locl.h:
```diff
@@ -108,6 +113,26 @@ struct evp_keyexch_st {
+    OSSL_OP_signature_verify_recover_fn *verify_recover;
+    OSSL_OP_signature_freectx_fn *freectx;
+    OSSL_OP_signature_dupctx_fn *dupctx;
+    OSSL_OP_signature_set_params_fn *set_params;
```
settable_params also?

paulidale replied:
I think so, even if it is always absent.

levitte replied:
Agreed, the type will need to exist

mattcaswell replied:
Yes - we do need settable_params. Now added.

slontis commented on crypto/evp/pmeth_fn.c:
```diff
@@ -13,105 +13,450 @@
 #include "evp_locl.h"
 
-int EVP_PKEY_sign_init(EVP_PKEY_CTX *ctx)
+static EVP_SIGNATURE *evp_signature_new(OSSL_PROVIDER *prov)
```
Looks like we will end up with quite a few of these near identical factory methods...
Could possibly be refactored at some point..

levitte replied:
I agree to a point...  there's a point where "refactoring" becomes a pile of magic that's hard to follow.  Of course *documenting the magic* (which we are still seriously bad at) might help making it less magic.

slontis commented on crypto/evp/pmeth_fn.c:
```diff
@@ -13,105 +13,450 @@
+                break;
+            signature->set_params = OSSL_get_OP_signature_set_params(fns);
+            break;
+        }
```
+ SETTABLE_PARAMS?

slontis commented on crypto/evp/pmeth_fn.c:
```diff
@@ -13,105 +13,450 @@
+    keymgmt_data.properties = properties;
+    return evp_generic_fetch(ctx, OSSL_OP_SIGNATURE, algorithm, properties,
+                             evp_signature_from_dispatch, &keymgmt_data,
+                             (int (*)(void *))EVP_SIGNATURE_up_ref,
```
Should the cast be extracted into a typdef?

mattcaswell replied:
I'm not sure how much that would really help readability here?

levitte replied:
The other variant is to have a few static functions that "translate" between `void *`, and `EVP_SIGNATURE *`.  That's how I do it in more of the code I've written, as that at least ensures some level of compatibility...  that being said, I'm fine with this sort of cast too...

richsalz replied:
Almost all other "callback" functions in OpenSSL API's have typedef's.  This one should have also had a typedef.

slontis commented on doc/man7/provider-signature.pod:
```diff
@@ -0,0 +1,229 @@
+
+ /* Signature parameters */
+ int OP_signature_set_params(void *ctx, const OSSL_PARAM params[]);
+
```
Settable_params?

slontis commented on doc/man7/provider-signature.pod:
```diff
@@ -0,0 +1,229 @@
+structure for holding context information during a signature operation.
+A pointer to this context will be passed back in a number of the other signature
+operation function calls.
+The paramater B<provctx> is the provider context generated during provider
```
Need to find all parameter references in this doc if they were copied from elsewhere (since they are wrong in that doc also). One of my PR's fixes it in one place..

mattcaswell replied:
I've fixed all the instances I could find in this doc.

slontis commented on doc/man7/provider-signature.pod:
```diff
@@ -0,0 +1,229 @@
+=head2 Signing Functions
+
+OP_signature_sign_init() initialises a context for signing given a provider side
+signature context in the B<ctx> paramter, and a pointer to a provider key object
```
parameter

slontis commented on doc/man7/provider-signature.pod:
```diff
@@ -0,0 +1,229 @@
+=head2 Verify Functions
+
+OP_signature_verify_init() initialises a context for verifying a signature given
+a provider side signature context in the B<ctx> paramter, and a pointer to a
```
param

paulidale replied:
_parameter_

mattcaswell replied:
I think there's something wrong with my keyboard. It doesn't seem to be able to write parameter correctly.

levitte replied:
Oh right, blame the keyboard :-D

slontis commented on doc/man7/provider-signature.pod:
```diff
@@ -0,0 +1,229 @@
+=head2 Verify Recover Functions
+
+OP_signature_verify_recover_init() initialises a context for recovering the
+signed data given a provider side signature context in the B<ctx> paramter, and
```
param

paulidale replied:
_parameter_

levitte commented on crypto/cms/cms_sd.c:
```diff
@@ -706,11 +706,23 @@ int CMS_SignerInfo_sign(CMS_SignerInfo *si)
+     * signer info data as a set of params...but that is non-trivial and since
+     * this isn't used by anything it may be better just to remove it.
+     */
+#if 0
```
Knee jerk reaction to `#if 0`?  I find the comment good enough explanation for not deleting the code until we have decided what to do with it.

paulidale replied:
Fair enough.

levitte commented on crypto/cms/cms_sd.c:
```diff
@@ -727,11 +739,23 @@ int CMS_SignerInfo_sign(CMS_SignerInfo *si)
+     * signer info data as a set of params...but that is non-trivial and since
+     * this isn't used by anything it may be better just to remove it.
+     */
+#if 0
```
Dito :wink:

levitte commented on crypto/dh/dh_ameth.c:
```diff
@@ -563,13 +563,13 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
     ossl_param_bld_init(&tmpl);
-    if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DH_P, p)
-        || !ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DH_G, g)
+    if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_FFC_P, p)
```
Uhmm, are we?  `EVP_PKEY` remains a container for the actual *key*, and these parameters are associate with the *key*, not with the operation.  However, if you absolutely want to associate them with an operation, then `KEYMGMT` would be the choice...

mattcaswell replied:
As @levitte said, I don't think we are trying to move away from PKEY, and these parameters are relevant to the *key* not the operation, which is why I use that in the name. We could lose the "P" and just make them `OSSL_KEY_PARAM_*` I suppose.

levitte replied:
I would rather keep the `P` to avoid confusion with symmetric keys

mattcaswell replied:
I've left this as is.

levitte commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -63,6 +65,7 @@ static int dsa_pub_decode(EVP_PKEY *pkey, X509_PUBKEY *pubkey)
         goto err;
     }
 
+    dsa->dirty_cnt++;
```
If more than one thread touches this at the same time, it will be changed either way, right?  At least one increment will happen, and that is enough to make this structure "dirty", because the key transferring mechanism looks for dirty count equality only.

paulidale replied:
Good point.

levitte commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -514,6 +521,51 @@ static int dsa_pkey_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
+    void *provkey = NULL;
+
+    if (p == NULL || q == NULL || g == NULL)
+        return NULL;
```
You need to remove the check of `pub_key` in `dh_pkey_export_to` as well.  And just so people don't start asking why we don't check for it, it might be good to have a comment both here and in `dh_pkey_export_to` that the key structure might only hold domain parameters...

mattcaswell replied:
Done

levitte commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -570,5 +622,13 @@ const EVP_PKEY_ASN1_METHOD dsa_asn1_meths[5] = {
-     old_dsa_priv_encode}
+     old_dsa_priv_encode,
+
+     NULL, NULL, NULL,
```
What would you like to see there instead?  These are ten method functions that aren't implemented in DSA, but are elsewhere, so I don't quite see how you'd avoid that amount of NULLs

levitte commented on crypto/evp/pmeth_lib.c:
```diff
@@ -422,6 +430,38 @@ int EVP_PKEY_CTX_set_dh_pad(EVP_PKEY_CTX *ctx, int pad)
+                            * be safe
+                            */
+                           (char *)name,
+                           strlen(name) + 1);
```
>  The construct calls add one internally.

Nope.  The `_set_` call does, yes.

paulidale replied:
My mistake.  I'm confusing the two.

levitte commented on crypto/evp/pmeth_lib.c:
```diff
@@ -422,6 +430,38 @@ int EVP_PKEY_CTX_set_dh_pad(EVP_PKEY_CTX *ctx, int pad)
+
+    mdsize = EVP_MD_size(md);
+    name = EVP_MD_name(md);
+    sig_md_params[0] = OSSL_PARAM_construct_utf8_string(
```
> I think I'll put in a PR where this call `OSSL_PARAM_construct_utf8_string` does a `strlen(3)` of its argument if the specified length is zero.

Yes, please do.

paulidale replied:
See #9760.  Every instance of `OSSL_PARAM_construct_utf8_string` now has a zero length.

Is it worth dropping this argument entirely?  I'm less sure about this change.

levitte replied:
No, because the construct function is also used to give a buffer where data should end up, by ways of `get_params` / `get_ctx_params`.  So the caller must be able to specify the buffer size.

levitte commented on crypto/include/internal/evp_int.h:
```diff
@@ -21,6 +21,9 @@ struct evp_pkey_ctx_st {
 
+    EVP_SIGNATURE *signature;
+    void *sigprovctx;
+
```
This turns out interesting...  can an `EVP_PKEY_CTX` be used for multiple operations (i.e. is the `operation` field a bit field rather than a number)?  If not, we might want to consider a union here...

mattcaswell replied:
No - I think you can only do one operation at a time.

levitte replied:
That's what I though...

mattcaswell replied:
I changed this to a union - although it turned out to be quite complicated.

levitte replied:
I can imagine.  Suddenly, you become *much* more dependent on the `operation` field.  Feel free to declare that not worth the while if you prefer...

levitte replied:
... oh, and having to create a small struct for each union variant, of course...!

mattcaswell replied:
Please take a look at feda77d and see what you think. I expect that there will be at least one more member in this union moving forward (possibly 2).

levitte commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -161,6 +161,17 @@ up to the size of the DH prime B<p>.
 
+=item OSSL_SIGNATURE_PARAM_MD_NAME (UTF8 string type)
+
+Sets the name of the digest algorithm used for the input to the signature
```
Regarding the item line, I think we should show the actual name (i.e. the name string).  I'm going to insist on it, actually.

mattcaswell replied:
> Regarding the item line, I think we should show the actual name (i.e. the name string). I'm going to insist on it, actually.

Instead of, or as well as the define?

levitte replied:
If you ask me, I would prefer the string to be the primary thing, but I'm absolutely with having existing macros displayed along with it, just as long as we don't make it look like the macro is the *only* possibility

paulidale replied:
The KDFs and MACs use the form:

```
=item B<OSSL_MAC_PARAM_DIGEST> ("digest") <utf8 string>
```

paulidale replied:
There has been a suggestion to change this to:
```
=item  "digest" B<OSSL_MAC_PARAM_DIGEST> <utf8 string>
```

levitte replied:
Or `=item  "digest" (B<OSSL_MAC_PARAM_DIGEST>) <utf8 string>`
(i.e. parenthesis added)

I would prefer this, i.e. emphasis on the string.

mattcaswell replied:
Is bare angled brackets around "utf8 string" correct pod syntax? It looks odd to me.

levitte replied:
It's perfectly acceptable pod syntax.  If you read [the perlpod manual](https://perldoc.perl.org/perlpod.html) carefully, you will see that the only real reason to use constructs like `E<lt>` and `E<gt>` is to disambiguate ambiguous cases, and that there are other ways that avoids that kind of construct.

(I actually asked myself the exact same question a few days ago, which is the reason I know this quite well now)

mattcaswell replied:
I made the edits as suggested by both @paulidale and @levitte.

levitte commented on include/openssl/core_names.h:
```diff
@@ -93,18 +93,27 @@ extern "C" {
 
+/* Signature parameters */
+#define OSSL_SIGNATURE_PARAM_MD_NAME    "signature-md-name"
+#define OSSL_SIGNATURE_PARAM_MD_SIZE    "signature-md-size"
```
Drop `signature-`

paulidale replied:
Are we using "digest" for the MD name?  At least, `OSSL_SIGNATURE_PARAM_DIGEST` even if the string doesn't change.

levitte replied:
This is a good point.  Also, `-name` is not necessary, `"digest"` is assumed to have a name as value, see `OSSL_MAC_PARAM_DIGEST` above

mattcaswell replied:
So should it be something like:
````
#define OSSL_SIGNATURE_PARAM_DIGEST         "digest"
#define OSSL_SIGNATURE_PARAM_DIGEST_SIZE    "digest-size"
````

paulidale replied:
I think so.

levitte replied:
Yes

mattcaswell replied:
Done

mattcaswell commented on crypto/dsa/dsa_ameth.c:
```diff
@@ -514,6 +521,51 @@ static int dsa_pkey_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
+                                    pub_key))
+            return NULL;
+
+        if (priv_key != NULL) {
```
I don't think private without public is allowed?

levitte replied:
Depends who you ask.  If you look at `DSA_set0_key()`, it looks permitted.  If you have a look at the `DSAPrivateKey` ASN.1 template, the public key must always be present (not optional).  if you look in `crypto/dsa/dsa_meth.c`'s `dsa_priv_decode`, you can see that a PKCS#8 structure only contains the private key plus domain parameters (`dsa_priv_decode` recalculates the public key).

Generally speaking, quite a lot of the libcrypto code assumes that an `EVP_PKEY` contains the public part.  That would probably be the only compelling reason to continue to make that mandatory.  However, I'm unsure if we should make that mandatory for providers.  After all, it's still possible to recalculate the public key from the domain parameters if it's missing, just like `dsa_priv_decode` does...

mattcaswell replied:
> If you look at DSA_set0_key(), it looks permitted.

No - it is not permitted there:

````
    if (d->pub_key == NULL && pub_key == NULL)
        return 0;
````

levitte replied:
Oh!  Sorry, I looked at `DH_set0_key()`, which doesn't have this check.  I wonder why they are different in this respect???

slontis commented on include/openssl/core_names.h:
```diff
@@ -93,18 +93,27 @@ extern "C" {
+#define OSSL_PKEY_PARAM_FFC_Q        "ffc-q"
+
 /* Diffie-Hellman Keys */
 #define OSSL_PKEY_PARAM_DH_PUB_KEY   "dh-pub"
```
Do we need individual names for pub, priv or could these just be common to all keys?

levitte replied:
Not for *all* keys.  RSA has n, e and d...

slontis replied:
you know what I meant :)

levitte replied:
Not sure.  One could interpret it to mean that all keys could share the same set of parameters, and simply pick and choose between them, i.e. something like this:

``` C
# define OSSL_PKEY_PARAM_PUB_KEY        "pub"
# define OSSL_PKEY_PARAM_PRIV_KEY       "priv"
# define OSSL_PKEY_PARAM_RSA_N          "n"
# define OSSL_PKEY_PARAM_RSA_E          "e"
# define OSSL_PKEY_PARAM_RSA_D          "d"
```

This is similar to how parameter names and macros are treated in the KDF port...

mattcaswell replied:
I wasn't sure what the conclusion was here, so I've not made any changes as yet.

levitte replied:
In that case, we should drop `ffc-` too.  I personally do not mind the shorter variants at all, quite the contrary.  If given via, say, `-pkeyopt` in one of the `pkey` sub-commands, I really cannot see a reason to be more verbose than `-pkeyopt hexn:xxxxxxxxxx`, say...

levitte replied:
BTW, did you notice @slontis' thumbs up on my comment from yesterday?  I would interpret that as a conclusion of sorts...  good enough for me!

levitte replied:
*Do* note that we're only talking about the strings here.  The macro names are fine, in my opinion.

mattcaswell commented on crypto/evp/pmeth_lib.c:
```diff
@@ -422,6 +430,38 @@ int EVP_PKEY_CTX_set_dh_pad(EVP_PKEY_CTX *ctx, int pad)
 
+int EVP_PKEY_CTX_set_signature_md(EVP_PKEY_CTX *ctx, const EVP_MD *md)
+{
+    OSSL_PARAM sig_md_params[3];
```
Can you give me a reference to where you've done that? I can't seem to find one.

levitte replied:
You need to look in @paulidale's KDF port.  But essentially, it allows this:

``` C
     *p++ = OSSL_PARAM_construct_utf8_string(
                           OSSL_SIGNATURE_PARAM_MD_NAME,
                           /*
                            * Cast away the const. This is read only so should
                            * be safe
                            */
                           (char *)name,
                           strlen(name) + 1);
    *p++ = OSSL_PARAM_construct_size_t(OSSL_SIGNATURE_PARAM_MD_SIZE, &mdsize);
    *p++ = OSSL_PARAM_construct_end();
```

... instead of

``` C
     sig_md_params[0] = OSSL_PARAM_construct_utf8_string(
                           OSSL_SIGNATURE_PARAM_MD_NAME,
                           /*
                            * Cast away the const. This is read only so should
                            * be safe
                            */
                           (char *)name,
                           strlen(name) + 1);
    sig_md_params[1] = OSSL_PARAM_construct_size_t(OSSL_SIGNATURE_PARAM_MD_SIZE,
                                                   &mdsize);
    sig_md_params[2] = OSSL_PARAM_construct_end();
```

mattcaswell replied:
I've used this construction now.

mattcaswell commented on crypto/evp/pmeth_lib.c:
```diff
@@ -422,6 +430,38 @@ int EVP_PKEY_CTX_set_dh_pad(EVP_PKEY_CTX *ctx, int pad)
+        return EVP_PKEY_CTX_ctrl(ctx, -1, EVP_PKEY_OP_TYPE_SIG,
+                                 EVP_PKEY_CTRL_MD, 0, (void *)(md));
+
+    if (md == NULL)
```
I actually changed this code around a little.

mattcaswell commented:
I've updated this with all the feedback received so far. I also rebased to resolve conflicts with master. Please take another look.

levitte commented on doc/man3/EVP_PKEY_CTX_ctrl.pod:
```diff
@@ -144,25 +150,50 @@ EVP_PKEY_CTX_set1_id, EVP_PKEY_CTX_get1_id, EVP_PKEY_CTX_get1_id_len
 =over 4
 
-=item OSSL_EXCHANGE_PARAM_PAD (uint type)
+=item "exchange-pad" (B<OSSL_EXCHANGE_PARAM_PAD>) <uint>
```
Thank you :-)

levitte commented:
There only seems to be the question about some parameter strings.  Otherwise, I believe this is pretty much done (at least with today's scrutiny)

mattcaswell commented:
I pushed an additional commit, changing the strings in core_names.h to use the "simple" version.

levitte commented:
Travis reports some memory leaks...  (this is a good opportunity to exercise asan / lsan, btw :wink:)

mattcaswell commented:
Fix up pushed addressing the memory leaks. Please take another look.

levitte commented:
> Fix up pushed addressing the memory leaks. Please take another look.

Oh, it was that simple?  You came off lucky

levitte commented on test/evp_extra_test.c:
```diff
@@ -1393,6 +1394,117 @@ static int test_EVP_CIPHER_fetch(int tst)
+     * to actually be *valid* parameters. We just need to set something. We
+     * don't even need a pub_key/priv_key.
+     */
+    dsa = DSA_new();
```
You need to include `opensl/dsa.h`

mattcaswell commented:
The CIs failed due to the inclusion issue you noted. I've pushed a fixup to see if it makes them happier. Assuming it does I'll take your approval and push to master.

levitte commented:
> The CIs failed due to the inclusion issue you noted.

That's why I noticed.  Otherwise, it would have flown right past me...

mattcaswell commented:
I resolved some trivial merge conflicts with master and then pushed. Thanks!
